### PR TITLE
7-Improve static context

### PR DESCRIPTION
**Pull request by CanBerker**
- Introduce InScopeVariable to static Context
- Cleanup Cosmetics of RuntimeIteratorVisitor
- Fix buggy _isRDD usage of HybridRuntimeIterator

### CODE REVIEW

File: src/main/java/sparksoniq/semantics/visitor/RuntimeIteratorVisitor.java [hunk 1]
```diff
@@ -20,7 +20,6 @@
 
 package sparksoniq.semantics.visitor;
```
Comment (CanBerker) [inline src/main/java/sparksoniq/semantics/visitor/RuntimeIteratorVisitor.java]: This file has only simple formatting changes.

File: src/main/java/sparksoniq/jsoniq/runtime/iterator/HybridRuntimeIterator.java [hunk 2]
```diff
@@ -106,7 +106,7 @@ public boolean hasNext() {
 
     @Override
     public Item next() {
-        if (!_isRDD) {
+        if (!isRDD()) {
             return nextLocal();
```
Comment (CanBerker) [inline src/main/java/sparksoniq/jsoniq/runtime/iterator/HybridRuntimeIterator.java]: Fix isRDD usage

File: src/main/java/sparksoniq/jsoniq/runtime/iterator/CommaExpressionIterator.java [hunk 3]
```diff
@@ -75,10 +75,6 @@ public void setNextResult() {
             }
         }
 
-        if (_nextResult == null) {
-            this._hasNext = false;
-        } else {
-            this._hasNext = true;
-        }
```
Comment (CanBerker) [inline src/main/java/sparksoniq/jsoniq/runtime/iterator/CommaExpressionIterator.java]: Simplify hasNext usage

File: src/main/java/sparksoniq/jsoniq/runtime/iterator/control/SwitchRuntimeIterator.java [hunk 4]
```diff
@@ -48,10 +48,8 @@ public SwitchRuntimeIterator(
     ) {
         super(null, iteratorMetadata);
         this._children.add(test);
-        for (RuntimeIterator key : cases.keySet())
-            this._children.add(key);
-        for (RuntimeIterator value : cases.values())
-            this._children.add(value);
+        this._children.addAll(cases.keySet());
+        this._children.addAll(cases.values());
         this._children.add(defaultReturn);
```
Comment (CanBerker) [inline src/main/java/sparksoniq/jsoniq/runtime/iterator/control/SwitchRuntimeIterator.java]: Simplify adding items to list

File: src/main/java/sparksoniq/semantics/StaticContext.java [hunk 5]
```diff
@@ -61,20 +84,34 @@ public boolean isInScope(String varName) {
...
+    private InScopeVariable getInScopeVariable(String varName) {
         if (_inScopeVariables.containsKey(varName))
-            return _inScopeVariables.get(varName).getFirst();
-        else if (_parent != null)
-            return _parent.getVariableSequenceType(varName);
-        else
+            return _inScopeVariables.get(varName);
+        else {
+            StaticContext ancestor = _parent;
+            while (ancestor != null) {
```
Comment (ghislainfourny) [inline src/main/java/sparksoniq/semantics/StaticContext.java]: why not a recursive call?

File: src/main/java/sparksoniq/semantics/StaticContext.java [hunk 6]
```diff
@@ -61,20 +84,34 @@ public boolean isInScope(String varName) {
...
+                if (ancestor._inScopeVariables.containsKey(varName)) {
+                    return ancestor._inScopeVariables.get(varName);
+                }
+                ancestor = ancestor.getParent();
```
Comment (ghislainfourny) [inline src/main/java/sparksoniq/semantics/StaticContext.java]: _parent

Comment (CanBerker) [reply src/main/java/sparksoniq/semantics/StaticContext.java]: Done

**Pull request merged**